Add dedicated feature flag for oauth device grant flow (#23891)

thomasdarimont opened the pull request:
The new flag "DEVICE_FLOW" is enabled by default for now. At a later stage we might disable the feature by default.

Users can now disable the device flow support for the whole server via: -Dkeycloak.profile.feature.device_flow=disabled
or
--features-disabled=device_flow

See discussion: https://github.com/keycloak/keycloak/discussions/23700

Fixes #23891

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Add dedicated feature flag for oauth device grant flow (#23891)

The new flag "DEVICE_FLOW" is enabled by default for now. At a later
stage we might disable the feature by default.

Users can now disable the device flow support for the whole server via:
-Dkeycloak.profile.feature.device_flow=disabled
or
--features-disabled=device_flow

See discussion: https://github.com/keycloak/keycloak/discussions/23700

Fixes #23891

thomasdarimont committed:
Honor device flow feature in client capability config in admin ui

thomasdarimont commented:
The following screenshots show the new behaviour when the device-flow is disabled for the Keycloak Server:

Realm Info lists device-flow as disabled feature
![image](https://github.com/keycloak/keycloak/assets/314690/75130607-dca1-465e-aafc-b7ae70464680)

Tokens Tab in Realm Settings doesn't show device flow settings
![image](https://github.com/keycloak/keycloak/assets/314690/1360e755-ae91-49b3-8db5-5dceac590b22)

When creating a new client Capability config doesn't show the device flow 
![image](https://github.com/keycloak/keycloak/assets/314690/6c5dacf7-d23d-46c8-8736-48b1d78d367d)

When editing a client device flow is also not shown:
![image](https://github.com/keycloak/keycloak/assets/314690/cea77993-e508-4859-bcdb-6c91a8ce75dd)

/realms/{realm}/device shows 404 page
![image](https://github.com/keycloak/keycloak/assets/314690/119dc84d-70b7-413b-817f-106b174da394)

jonkoops commented on common/src/main/java/org/keycloak/common/Profile.java:
```diff
@@ -92,7 +92,12 @@ public enum Feature {
+
+        DEVICE_FLOW("OAuth 2.0 Device Authorization Grant", Type.DEFAULT),
+
+        // Add new Feature above this line
+        ;
```
I don't think we need to leave a comment for this, it should be pretty clear from the code itself where this goes.

thomasdarimont replied:
Thanks for the review. I removed the comment.

jonkoops reviewed (dismissed):
Changes to the front-end code look good to me. I will leave it to other team members to judge if this feature is something we want and if the Java code is adequate.

Michito-Okai reviewed (approved):
Looks good to me. Please insert its test code afterwards.

thomasdarimont committed:
Revise Profile based on code review

thomasdarimont committed:
Add tests for oauth device grant flow feature disabled (#23892)

thomasdarimont commented:
@mposolda @tnorimat I just added the missing tests and I think the feature is complete now. Feel free to merge if the pipeline passes.

thomasdarimont committed:
Revise tests for disabled device flow

thomasdarimont committed:
Add device-flow flag to approved help output references

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

mposolda commented:
@thomasdarimont Thanks for the PR and Thanks everyone for the review!

Looks the failing test in "Keycloak CI / Base IT (6)" is regression of this PR.
```
Error: OR] Failures: 
Error: OR] org.keycloak.testsuite.oauth.OAuth2DeviceAuthorizationGrantTest.ensureDeviceFlowConfigNotPresentWhenDeviceFlowDisabled
Error: OR]   Run 1: OAuth2DeviceAuthorizationGrantTest.ensureDeviceFlowConfigNotPresentWhenDeviceFlowDisabled:1019 Should return not found for device auth endpoint expected:<404> but was:<204>
Error: OR]   Run 2: OAuth2DeviceAuthorizationGrantTest.ensureDeviceFlowConfigNotPresentWhenDeviceFlowDisabled:1019 Should return not found for device auth endpoint expected:<404> but was:<204>
Error: OR]   Run 3: OAuth2DeviceAuthorizationGrantTest.ensureDeviceFlowConfigNotPresentWhenDeviceFlowDisabled:1019 Should return not found for device auth endpoint expected:<404> but was:<204>
-- [INFO] 
-- [INFO] 
Error: OR] Tests run: 1149, Failures: 1, Errors: 0, Skipped: 22
```
Could you please take a look?

thomasdarimont commented:
Thanks for taking a look! What's different in "Keycloak CI / Base IT (6)"?  (Perhaps a better name would really help here)
Locally it works fine, I try to replicate whatever "Keycloak CI / Base IT (6)" later today.

thomasdarimont committed:
Revise approved help output tests for disabled device flow

thomasdarimont committed:
Revise test method name

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

mposolda commented:
> Thanks for taking a look! What's different in "Keycloak CI / Base IT (6)"? (Perhaps a better name would really help here) Locally it works fine, I try to replicate whatever "Keycloak CI / Base IT (6)" later today.

@thomasdarimont It uses keycloak distribution instead of the undertow distribution. Maybe you can replicate the issue when you run the test with something like:
```
mvn clean install -DskipTests=true
cd testsuite/integration-arquillian
mvn clean install -Pauth-server-quarkus -Dtest=OAuth2DeviceAuthorizationGrantTest
```
?

Just a guess (did not tried it myself).

thomasdarimont commented:
@mposolda I gave this another try, and it seems as if the `@DisableFeature(value = Profile.Feature.DEVICE_FLOW, skipRestart = true)` is not taken into account by the "Keycloak CI / Base IT (6)" testrun.

When I run the tests for ` OAuth2DeviceAuthorizationGrantTest` in the IDE, it works fine, but fails when run with 
` mvn clean install -Pauth-server-quarkus -Dtest=OAuth2DeviceAuthorizationGrantTest` -rf :integration-arquillian-tests-base

Could it be that the undertow based test-server implementation doesn't consider the feature flags?

thomasdarimont commented:
@mposolda the problem seems to be that the test is run with  `$KEYCLOAK_PROJECT_HOME/testsuite/integration-arquillian/tests/base/target/containers/auth-server-undertow/profile.properties` 
which for some reason does not exist when I run only the tests from the `OAuth2DeviceAuthorizationGrantTest`... 
![image](https://github.com/keycloak/keycloak/assets/314690/65361aa7-223c-4a0a-94c7-4a65bf0d3aab)

thomasdarimont committed:
Activate device_flow feature by default for OAuth2DeviceAuthorizationGrantTest

thomasdarimont committed:
Fix typo in TestingResourceProvider#enableFeature

thomasdarimont commented:
I could successfully run the build via:
````
mvn clean install -Pauth-server-undertow -Dtest=OAuth2DeviceAuthorizationGrantTest --am
````

thomasdarimont committed:
Workaround non working feature toggle with -Pauth-server-quarkus

For some reason @DisableFeature is not considered when executed with -Pauth-server-quarkus.

Current workaround is to ignore exceptions during -Pauth-server-quarkus.

thomasdarimont commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OAuth2DeviceAuthorizationGrantTest.java:
```diff
@@ -992,6 +997,43 @@ public void testClientWithErrors() throws Exception {
+    public void ensureDeviceFlowConfigNotPresentWhenDeviceFlowIsDisabled() throws Exception {
+
+        // this test currently does not work with -Pauth-server-quarkus
+        ContainerAssume.assumeAuthServerUndertow();
```
After many hours of trying to making this test work with -Pauth-server-quarkus I gave up and just assume that we run the test with undertow. The test works fine in the IDE and with -Pauth-server-undertow.

For some reason the @DisabledFeature annotation is now applied when the test suite is executed with -Pauth-server-quarkus.

mposolda replied:
That is fine, Thanks a lot for this investigation!

Created follow-up issue for looking at this test https://github.com/keycloak/keycloak/issues/24241. When thinking about this, I wonder if it can have some relation to OIDC well-known not being refreshed after feature is enabled/disabled? But not sure...

savage-alex commented:
Thank you for this fix
